internal/controllers/reference/reference.go

```diff
@@ -0,0 +1,150 @@
+package reference
+
+import (
+	"context"
+
+	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/kong/kubernetes-ingress-controller/v2/internal/dataplane"
+)
+
+const (
+	VersionV1      = "v1"
+	KindSecret     = "Secret"
+	CACertLabelKey = "konghq.com/ca-cert"
+)
+
+// UpdateReferencesToSecret updates the reference records between referrer and each secret
+// in namespacedNames in record cache.
+func UpdateReferencesToSecret(
+	ctx context.Context,
+	c client.Client, indexers CacheIndexers, dataplaneClient *dataplane.KongClient,
+	referrer client.Object, namespacedNames []types.NamespacedName,
+) error {
+	for _, nsName := range namespacedNames {
+
+		secret := &corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: nsName.Namespace,
+				Name:      nsName.Name,
+			},
+		}
+
+		// Here we update the reference relationship even when the referred secret does not exist yet
+		// If the referred secret is created, it could be reconciled in secret controller.
+		referrerCopy := referrer.DeepCopyObject().(client.Object)
+		if err := indexers.SetObjectReference(
+			referrerCopy, secret.DeepCopy()); err != nil {
+			return err
+		}
+
+		if err := c.Get(ctx, nsName, secret); err != nil {
+			return err
+		}
+
+		if err := dataplaneClient.UpdateObject(secret); err != nil {
+			return err
+		}
+	}
+
+	return removeOutdatedReferencesToSecret(ctx, indexers, c, dataplaneClient, referrer, namespacedNames)
+}
+
+// removeOutdatedReferenceToSecret removes outdated reference records to secrets in reference indexer.
+// secrets that are really referred by referrer is passed in parameter referredSecretNames.
+// if the secret is not actually referrenced by any object after deleting outdated reference records,
+// the secret will be removed from object cache in dataplaneClient.
+func removeOutdatedReferencesToSecret(
+	ctx context.Context,
+	indexers CacheIndexers, c client.Client, dataplaneClient *dataplane.KongClient,
+	referrer client.Object, referredSecretNames []types.NamespacedName,
+) error {
+	referredSecretNameMap := make(map[types.NamespacedName]bool, len(referredSecretNames))
+	for _, nsName := range referredSecretNames {
+		referredSecretNameMap[nsName] = true
+	}
```

Comment:
- Since we use this map as a set and we only care about whether an object is in the map, not what is its value, let's use a `struct{}` as map's value type.

  ```suggestion
	  referredSecretNameMap := make(map[types.NamespacedName]struct{}, len(referredSecretNames))
	  for _, nsName := range referredSecretNames {
		  referredSecretNameMap[nsName] = struct{}{}
	  }
  ```
  
- It just occurred to me that this will cause us to **always** allocated a map and copy objects form the slice of namespaced names into this map just for lookup purposes. These are basically created in `list***ReferredSecrets` functions. It might be worth considering changing the original implementation which creates those slices into maps so that we don't need to create a lookup map every time here.

Reply:
Good idea, but quite a lot of code changes. Will submit the change after a while.

Reply:
updated.